### PR TITLE
[AOTI] Implement support for user defined kernels that use triton.autotune

### PR DESCRIPTION
**Pull request by oulgen**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #113317
* __->__ #113229

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

### CODE REVIEW

File: torch/_inductor/codegen/wrapper.py [hunk 1]
```diff
@@ -1754,13 +1754,22 @@ def generate_extern_kernel_out(self, output_view, codegen_reference, args, kerne
...
+        assert len(grid) != 0
+        if len(grid) == 1:
+            grid_decision = grid[0]
+        else:
+            meta = CudaKernelParamCache.get(kernel_name)
+            assert meta is not None
+            grid_decision = None
+            for i, c in enumerate(configs):
+                if all(arg == meta[key] for key, arg in c.kwargs.items()):
+                    grid_decision = grid[i]
```
Comment (chenyang78) [inline torch/_inductor/codegen/wrapper.py:1766]: Do we want to break from the loop if we get one valid choice?

**Review by chenyang78 — APPROVED** [review]
Left a minor comment. LGTM overall. Thanks!

File: torch/_inductor/triton_heuristics.py [hunk 2]
```diff
@@ -460,6 +460,10 @@ def save_cuda_kernel(self, grid, stream, launcher):
             "num_warps": launcher.bin.num_warps,
             "shared_mem": launcher.bin.shared,
             "stream": stream,
+            # User defined triton kernels will have arbitrary kwarg names
+            # This might also add XBLOCK, YBLOCK, ZBLOCK but that is since
+            # no one will read it.
+            **launcher.config.kwargs,
```
Comment (aakhundov) [inline torch/_inductor/triton_heuristics.py:466]: Can we have run into an edge case that user-defined param names coincide with some of the keys above? If so, should we put `launcher.config.kwargs` into a sub-dict?

File: torch/_inductor/codegen/wrapper.py (hunk 1 above)
Comment (aakhundov) [reply torch/_inductor/codegen/wrapper.py:1766]: Or, being defensive, `assert grid_decision is None` before the assignment.

**Review by aakhundov — COMMENTED** [review]
Neat, thanks @oulgen!

**Pull request closed without merging**